README.md

```diff
@@ -1,2 +1,195 @@
+3. The signature is encrypted using AES (variable).
+4. The encrypted signature is looked up in the database, failing validation if no such row exists.
+5. They key is considered valid and is now subject for other validations, like audience, redirect or state matching.
+
+A token generated by `generator.CryptoGenerator` looks like:
+
+```
+GUULhK6Od/7UAlnKvMau8APHSKXSRwm9aoOk56SHBns.JDJhJDEwJDdwVmpCQmJLYzM2VDg1VHJ5aEdVOE81NVdRSkt6bHBHR1QwOC9pbTNFWmpQRXliTWRPeDQy
+```
+
+#### Encrypt credentials at rest
+
+Credentials (token signatures, passwords and secrets) are always encrypted at rest.
+
+#### Implement peer reviewed IETF Standards
+
+Fosite implements [rfc6749](https://tools.ietf.org/html/rfc6749) and enforces countermeasures suggested in [rfc6819](https://tools.ietf.org/html/rfc6819).
+
+### Provide extensibility and interoperability
+
+... because OAuth2 is an extensible and flexible **framework**. Fosite let's you register new response types, new grant
+types and new response key value pares. This is useful, if you want to provide OpenID Connect on top of your
+OAuth2 stack. Or custom assertions, what ever you like and as long as it is secure. ;)
+
+## Usage
+
+This section is WIP and we welcome discussions via PRs or in the issues.
+
+### Store
+
+To use fosite, you need to implement `fosite.Storage`. Example implementations (e.g. postgres) of `fosite.Storage`
+will be added in the close future.
+
+### Authorize Endpoint
+
+```go
+package main
+
+import "github.com/ory-am/fosite"
+import "github.com/ory-am/fosite/session"
+import "github.com/ory-am/fosite/storage"
+
+// Let's assume that we're in a http handler
+func handleAuth(rw http.ResponseWriter, req *http.Request) {
+    oauth2 := fosite.NewDefaultOAuth2()
+
+    // ** This part of the API is not finalized yet **
+    // var store storage.Storage
+    // this should be something like: store := fosite.NewPostgreSQLStore()
+    // ****
+
+    authorizeRequest, err := oauth2.NewAuthorizeRequest(r, store)
+    if err != nil {
+        // ** This part of the API is not finalized yet **
+        // oauth2.RedirectError(rw, error)
+        // oauth2.WriteError(rw, error)
+        // oauth2.handleError...
+    // ****
+        return
+    }
+
+    // you have now access to authorizeRequest, Code ResponseTypes, Scopes ...
+    // and can show the user agent a login or consent page.
+
+    // it would also be possible to redirect the user to an identity provider (google, microsoft live, ...) here
+    // and do fancy stuff like OpenID Connect amongst others
+
+    // Once you have confirmed the users identity and consent that he indeed wants to give app XYZ authorization,
+    // you will use the user's id to create an authorize session
+    user := "12345"
+    session := session.NewAuthorizeSession(authorizeRequest, user)
+
+    // You can store additional metadata, for example
+    // session.SetExtra(map[string]interface{}{
```

Comment:
Hey I just saw that SetExtra is defined to use interface{}, not map[string]interface{}. I think this is smarter because the developer has complete freeness over what he passes to extra. What do you think? You can check the interface definition here: https://github.com/ory-am/fosite/blob/unstaged/session.go#L12